### ISSUE TITLE
AUTOCOMPLETION. ng completion has no output

### ISSUE THREAD

**Issue by MRokas** [angular/angular-cli#2253]
> ## Please provide us with the following information:

1. OS?  Windows 7, 8 or 10. Linux (which distribution). Mac OSX (Yosemite? El Capitan?)
Ubuntu 16 64bit

2. Versions. Please run `ng --version`. If there's nothing outputted, please run
in a Terminal: `node --version` and paste the result here:

```
angular-cli: 1.0.0-beta.14
node: 6.6.0
os: linux x64
```

3. Repro steps. Was this an app that wasn't created using the CLI? What change did you
   do on your code? etc.

```
ng completion >> ~/.bashrc
source ~/.bashrc
```

there was no new text inside /.bashrc
trying to use it on any other .txt file, has same result

4. The log given by the failure. Normally this include a stack trace and some
   more information.

5. Mention any other details that might be useful.

> ----------------------------------ng completion-----------------------------
> Thanks! We'll be in touch soon.

**Comment by filipesilva** [comment]
Does this still happen? @catull has made some nice improvement to `ng completion` that might have fixed your issue. I'm on windows myself so cannot repro.

**Comment by anisabboud** [comment]
~$ ng version
    angular-cli: 1.0.0-beta.16
    node: 6.7.0
    os: linux x64

    ~$ ng completion
    The specified command completion is invalid. For available options, see `ng help`.

**Comment by catull** [comment]
**TL;DR**: Apparently, ng now requires either a file or directory named .angular-cli.
It expects it either in $HOME/ or the current working directory.

However, I got the same result when $HOME/.angular-cli was a directory or an empty file.
Same case with ./.angular-cli as a directory or an empty file in the workspace (~/workspace/js/ng/angular-cli.git).

All invocations of `ng completion` were performed in that workspace.
All gave the same output below.
I only added 2 lines to a dependent module - yam - locally.
Even before I started adding .angular-cli in either directories, the error was the same.
Could be a [red herring.](http://www.merriam-webster.com/dictionary/red%20herring)


## The long story ...

I traced this one down to the following line:

/usr/local/lib/node_modules/angular-cli/node_modules/yam/lib/utils/io-utils.js:27

Lines 21ff are:

```javascript
var exists = function(path) {
  var stat;

  try {
    stat = fsExtra.statSync(path);
  } catch(e) {
    console.log("Path " + path); // ADDED THIS FOR BETTER DEBUGGING
    console.log(e);              // ADDED THIS ONE TOO
    return false;                // <- THIS IS LINE 27, THE TWO STATEMENTS ABOVE WERE ADDED
  }

  return stat;
};
```

The following command is typed into a zsh:

```shell
ng completion
```

The output is - user name redacted:
```log
iPath /Users/#######/.angular-cli
{ Error: ENOENT: no such file or directory, stat '/Users/#######/.angular-cli'
    at Error (native)
    at Object.fs.statSync (fs.js:987:18)
    at exists (/usr/local/lib/node_modules/angular-cli/node_modules/yam/lib/utils/io-utils.js:25:20)
    at isDirectory (/usr/local/lib/node_modules/angular-cli/node_modules/yam/lib/utils/io-utils.js:6:14)
    at new Config (/usr/local/lib/node_modules/angular-cli/node_modules/yam/lib/config.js:17:7)
    at new Yam (/usr/local/lib/node_modules/angular-cli/node_modules/yam/lib/yam.js:37:5)
    at module.exports (/usr/local/lib/node_modules/angular-cli/node_modules/angular-cli/lib/cli/index.js:50:16)
    at module.exports (/usr/local/lib/node_modules/angular-cli/lib/cli/index.js:45:10)
    at /usr/local/lib/node_modules/angular-cli/bin/ng:36:5
    at /usr/local/lib/node_modules/angular-cli/node_modules/resolve/lib/async.js:46:14
    at process (/usr/local/lib/node_modules/angular-cli/node_modules/resolve/lib/async.js:173:43)
    at ondir (/usr/local/lib/node_modules/angular-cli/node_modules/resolve/lib/async.js:188:17)
    at load (/usr/local/lib/node_modules/angular-cli/node_modules/resolve/lib/async.js:69:43)
    at onex (/usr/local/lib/node_modules/angular-cli/node_modules/resolve/lib/async.js:92:31)
    at /usr/local/lib/node_modules/angular-cli/node_modules/resolve/lib/async.js:22:47
    at FSReqWrap.oncomplete (fs.js:123:15)
  errno: -2,
  code: 'ENOENT',
  syscall: 'stat',
  path: '/Users/#######/.angular-cli' }
Path /Users/#######/.angular-cli
{ Error: ENOENT: no such file or directory, stat '/Users/#######/.angular-cli'
    at Error (native)
    at Object.fs.statSync (fs.js:987:18)
    at exists (/usr/local/lib/node_modules/angular-cli/node_modules/yam/lib/utils/io-utils.js:25:20)
    at isFile (/usr/local/lib/node_modules/angular-cli/node_modules/yam/lib/utils/io-utils.js:14:14)
    at new Config (/usr/local/lib/node_modules/angular-cli/node_modules/yam/lib/config.js:20:14)
    at new Yam (/usr/local/lib/node_modules/angular-cli/node_modules/yam/lib/yam.js:37:5)
    at module.exports (/usr/local/lib/node_modules/angular-cli/node_modules/angular-cli/lib/cli/index.js:50:16)
    at module.exports (/usr/local/lib/node_modules/angular-cli/lib/cli/index.js:45:10)
    at /usr/local/lib/node_modules/angular-cli/bin/ng:36:5
    at /usr/local/lib/node_modules/angular-cli/node_modules/resolve/lib/async.js:46:14
    at process (/usr/local/lib/node_modules/angular-cli/node_modules/resolve/lib/async.js:173:43)
    at ondir (/usr/local/lib/node_modules/angular-cli/node_modules/resolve/lib/async.js:188:17)
    at load (/usr/local/lib/node_modules/angular-cli/node_modules/resolve/lib/async.js:69:43)
    at onex (/usr/local/lib/node_modules/angular-cli/node_modules/resolve/lib/async.js:92:31)
    at /usr/local/lib/node_modules/angular-cli/node_modules/resolve/lib/async.js:22:47
    at FSReqWrap.oncomplete (fs.js:123:15)
  errno: -2,
  code: 'ENOENT',
  syscall: 'stat',
  path: '/Users/#######/.angular-cli' }
Path /Users/#######/workspace/js/ng/angular-cli.git/.angular-cli
{ Error: ENOENT: no such file or directory, stat '/Users/#######/workspace/js/ng/angular-cli.git/.angular-cli'
    at Error (native)
    at Object.fs.statSync (fs.js:987:18)
    at exists (/usr/local/lib/node_modules/angular-cli/node_modules/yam/lib/utils/io-utils.js:25:20)
    at isDirectory (/usr/local/lib/node_modules/angular-cli/node_modules/yam/lib/utils/io-utils.js:6:14)
    at new Config (/usr/local/lib/node_modules/angular-cli/node_modules/yam/lib/config.js:17:7)
    at new Yam (/usr/local/lib/node_modules/angular-cli/node_modules/yam/lib/yam.js:38:5)
    at module.exports (/usr/local/lib/node_modules/angular-cli/node_modules/angular-cli/lib/cli/index.js:50:16)
    at module.exports (/usr/local/lib/node_modules/angular-cli/lib/cli/index.js:45:10)
    at /usr/local/lib/node_modules/angular-cli/bin/ng:36:5
    at /usr/local/lib/node_modules/angular-cli/node_modules/resolve/lib/async.js:46:14
    at process (/usr/local/lib/node_modules/angular-cli/node_modules/resolve/lib/async.js:173:43)
    at ondir (/usr/local/lib/node_modules/angular-cli/node_modules/resolve/lib/async.js:188:17)
    at load (/usr/local/lib/node_modules/angular-cli/node_modules/resolve/lib/async.js:69:43)
    at onex (/usr/local/lib/node_modules/angular-cli/node_modules/resolve/lib/async.js:92:31)
    at /usr/local/lib/node_modules/angular-cli/node_modules/resolve/lib/async.js:22:47
    at FSReqWrap.oncomplete (fs.js:123:15)
  errno: -2,
  code: 'ENOENT',
  syscall: 'stat',
  path: '/Users/#######/workspace/js/ng/angular-cli.git/.angular-cli' }
Path /Users/#######/workspace/js/ng/angular-cli.git/.angular-cli
{ Error: ENOENT: no such file or directory, stat '/Users/#######/workspace/js/ng/angular-cli.git/.angular-cli'
    at Error (native)
    at Object.fs.statSync (fs.js:987:18)
    at exists (/usr/local/lib/node_modules/angular-cli/node_modules/yam/lib/utils/io-utils.js:25:20)
    at isFile (/usr/local/lib/node_modules/angular-cli/node_modules/yam/lib/utils/io-utils.js:14:14)
    at new Config (/usr/local/lib/node_modules/angular-cli/node_modules/yam/lib/config.js:20:14)
    at new Yam (/usr/local/lib/node_modules/angular-cli/node_modules/yam/lib/yam.js:38:5)
    at module.exports (/usr/local/lib/node_modules/angular-cli/node_modules/angular-cli/lib/cli/index.js:50:16)
    at module.exports (/usr/local/lib/node_modules/angular-cli/lib/cli/index.js:45:10)
    at /usr/local/lib/node_modules/angular-cli/bin/ng:36:5
    at /usr/local/lib/node_modules/angular-cli/node_modules/resolve/lib/async.js:46:14
    at process (/usr/local/lib/node_modules/angular-cli/node_modules/resolve/lib/async.js:173:43)
    at ondir (/usr/local/lib/node_modules/angular-cli/node_modules/resolve/lib/async.js:188:17)
    at load (/usr/local/lib/node_modules/angular-cli/node_modules/resolve/lib/async.js:69:43)
    at onex (/usr/local/lib/node_modules/angular-cli/node_modules/resolve/lib/async.js:92:31)
    at /usr/local/lib/node_modules/angular-cli/node_modules/resolve/lib/async.js:22:47
    at FSReqWrap.oncomplete (fs.js:123:15)
  errno: -2,
  code: 'ENOENT',
  syscall: 'stat',
  path: '/Users/#######/workspace/js/ng/angular-cli.git/.angular-cli' }
The specified command completion is invalid. For available options, see `ng help`.
```

**Comment by catull** [comment]
My environment:  macOS Sierra (10.12).
Shell: zsh 5.2.0
npm: 3.10.3

When issuing `ng --version; echo $?`, I also the the output above now, but in the end I get:
```log
angular-cli: 1.0.0-beta.16
node: 6.7.0
os: darwin x64
0
```
The error code is 0 though, unlike with `ng completion`.

**Comment by filipesilva** [comment]
That's not right. I'll look into it.

**Comment by catull** [comment]
By the way, a recently installed angular-cli from npm repository,
`ng version`
```log
angular-cli: 1.0.0-beta.17
node: 6.7.0
os: darwin x64
```
finally produces useful output:

```shell
###-begin-ng-completion###
#
# ng command completion script
#
# Installation: ng completion >> ~/.bashrc (or ~/.zshrc)
#

ng_opts='b build completion doc e2e g generate get github-pages:deploy gh-pages:deploy h help i init install lint make-this-awesome new s serve server set t test v version -h --help'

build_opts='--aot --base-href --environment --output-path --suppress-sizes --target --watch --watcher -bh -dev -e -o -prod -t -w'
generate_opts='class component directive enum module pipe route service c cl d e m p r s --help'
github_pages_deploy_opts='--base-href --environment --gh-token --gh-username --message --skip-build --target --user-page -bh -e -t'
help_opts='--json --verbose -v'
init_opts='--dry-run inline-style inline-template --link-cli --mobile --name --prefix --routing --skip-bower --skip-npm --source-dir --style --verbose -d -is -i

....
```


Good work!

**Comment by anisabboud** [comment]
Cool! Works for me too after updating the CLI. Thank you! This thread seems related to https://github.com/angular/angular-cli/pull/1890.

Running `ng completion >> ~/.bashrc` as per the instructions didn't append anything to my ~/.bashrc, so I manually copied and pasted the output of `ng completion` to my ~/.bashrc.

When running `source ~/.bashrc` I got a syntax error, perhaps because the two lines that end with `opts=$github_pages_deploy_opts` in the output of `ng completion` are missing `;;` in the end? Adding these solved it for me.

**Comment by catull** [comment]
@anisabboud Good catch, the semi-colons are missing. Created a new PR for this.

**Comment by anisabboud** [comment]
Thanks @catull! That was quick.

Any idea why `ng completion >> ~/.bashrc` doesn't work? (mentioned in https://github.com/angular/angular-cli/#commands-autocompletion)

`ng completion &>> ~/.bashrc` (with `&>>`) works for me, so it seems like `ng completion` is printing to `stderr` instead of `stdout`..

**Comment by catull** [comment]
Perhaps this should be the way to redirect standard output - not standand error diagnostic:
```bash
ng completion 1>> ~/.bashrc
```
I am not a bash user, but the statement would NOT append error output to ~/.bashrc.
Can you verify ?
This works in zsh:
```zsh
ng completion 1>>~/.zshrc
```
The '|' is necessary to create it freshly.


The above statement would work in zsh, though.

**Comment by anisabboud** [comment]
`ng completion 1>> ~/.bashrc` is not doing anything for me.

Looks like [console.log(scriptOutput)](https://github.com/angular/angular-cli/blob/0162f71446dec90f8ed49d07335432c9b49b8f62/packages/angular-cli/commands/completion.ts#L14) is printing to `stderr` even though the nodejs [documentation](https://nodejs.org/api/console.html#console_console_log_data) says it prints to `stdout`.

**Comment by catull** [comment]
Actually, if completion.sh printed to stderr, then the behaviour described initially is correct, because an error happened causing the script to be exited without useful output.

The original poster did not volunteer more information, so it is difficult to fix his/her situation.
It's all guess-work now.

**Comment by chpasha** [comment]
I have the same issue on ubuntu 14.04 - ng completion issues the script code, but >> does nothing

**Comment by catull** [comment]
Can you provide more pieces of information, please ?

Which shell, and which version of it ?
Which version of node, npm and ng do you use ?

**Comment by chpasha** [comment]
GNU bash, version 4.3.11(1)-release (x86_64-pc-linux-gnu)
Node v6.2.0
npm 3.8.9
angular-cli -1.0.0-beta.17

**Comment by anisabboud** [comment]
```
~$ lsb_release -a
Description:	Ubuntu 16.04.1 LTS

~$ bash --version
GNU bash, version 4.3.46(1)-release (x86_64-pc-linux-gnu)

~$ npm -v
3.10.3

~$ ng -v
angular-cli: 1.0.0-beta.17
node: 6.7.0
os: linux x64
```

To sum up my experience with `ng completion`:

- `ng completion` works and prints the output to the terminal.
- `ng completion >> ~/.bashrc` does nothing (which is the issue/bug).
- `ng completion &>> ~/.bashrc` works and appends stuff to my `~/.bashrc`.

**Comment by chpasha** [comment]
I have absolutely the same symptoms

**Comment by catull** [comment]
Embarassingly my situation is much similar.

OS: macOS Sierra (10.12)
npm: 3.10.8
angular-cli: 1.0.0-beta.17
node: 6.7.0
os: darwin x64
shell: zsh 5.2

In zsh, executing `ng completion` displays the shell script pieces.
But, all these attempts

- `ng completion >>someFile`
- `ng completion >>| someOtherFile`
- `ng completion 1>>anotherFile`
- `ng completion 2>>oneMoreFile`

produce empty files.

The solution is to use this frankensteinian monstruction:

`ng complete 1>>~/.bashrc 2>>&`

Will produce a new PR.

**Comment by catull** [comment]
Ready for merging with PR #2635.
Enjoy!
